src/program_generator/common.py

```diff
@@ -36,21 +36,38 @@ def generate_single_marginal(single_op_pushes, operation, op_count):
 
   interleaved_op_and_pops_count = max(op_count - 1, 0)
   end_pop_count = total_pop_count - interleaved_op_and_pops_count * nreturns
-    
-  empty_pushes = ["6000"] * empty_push_count
-  pushes = single_op_pushes * ceil(push_count / arity) if arity > 0 else []
 
-  if op_count == 0:
-    middle = []
-    pops = [popcode] * total_pop_count
-  elif op_count >= 1:
-    middle = [opcode] + ([popcode] * nreturns + [opcode]) * interleaved_op_and_pops_count
-    pops = [popcode] * end_pop_count
+  # start generation
+  bytecode = ''
 
   # If this is an OPCODE accessing memory, we pre-allocate 128KB of memory at the very beginning
-  initial_mstore = [initial_mstore_bytecode()] if operation['Mnemonic'] in constants.MEMORY_OPCODES else []
-
-  bytecode = ''.join(initial_mstore + empty_pushes + pushes + middle + pops)
+  initial_mstore = initial_mstore_bytecode() if operation['Mnemonic'] in constants.MEMORY_OPCODES else ''
+  bytecode += initial_mstore
+    
+  empty_pushes = ["6000"] * empty_push_count
+  bytecode += ''.join(empty_pushes)
+  pushes = single_op_pushes * ceil(push_count / arity(operation)) if arity(operation) > 0 else []
+  bytecode += ''.join(pushes)
+
+  if operation['Mnemonic'] in ["JUMP", "JUMPI"]:
+    # JUMPs don't return anything, we don't POP it, so assertion
+    assert nreturns == 0
+
+    empty_combos_count = MAX_INSTRUCTIONS - op_count
+    for _ in range(0, op_count):
+      bytecode += jump_opcode_combo(bytecode, opcode)
+    for _ in range(0, empty_combos_count):
+      bytecode += jump_opcode_combo(bytecode, None)
+  else:
+    if op_count == 0:
+      middle = []
+      pops = [popcode] * total_pop_count
+      bytecode += ''.join(middle + pops)
+    elif op_count >= 1:
+      middle = [opcode] + ([popcode] * nreturns + [opcode]) * interleaved_op_and_pops_count
+      bytecode += ''.join(middle)
+      pops = [popcode] * end_pop_count
+      bytecode += ''.join(pops)
```

Comment:
Note this bit - it should behave exactly as before the change, it's only emitting the bytecode straight into the `bytecode` variable, instead of building the arrays to compose later.